### PR TITLE
Update Sre Engine Implementing to CPython 3.12

### PR DESCRIPTION
**Pull request by qingshi163**
https://github.com/RustPython/sre-engine/pull/17

### CODE REVIEW

**Comment by youknowone** [comment]
@qingshi163 can `sre-engine` be used as an indepdendent crate out of RustPython? If not, how about putting it in this repository as `rustpython-sre-engine`?

**Comment by qingshi163** [comment]
> @qingshi163 can `sre-engine` be used as an indepdendent crate out of RustPython? If not, how about putting it in this repository as `rustpython-sre-engine`?

sure

File: vm/Cargo.toml [hunk 1]
```diff
@@ -78,9 +78,9 @@ result-like = "0.4.5"
 timsort = "0.1.2"
 
 # RustPython crates implementing functionality based on CPython
-sre-engine = "0.4.1"
+# sre-engine = "0.4.1"
 # to work on sre-engine locally or git version
-# sre-engine = { git = "https://github.com/RustPython/sre-engine", rev = "refs/pull/14/head" }
+sre-engine = { git = "https://github.com/RustPython/sre-engine", rev = "refs/pull/17/head" }
```
Comment (youknowone) [inline vm/Cargo.toml:83]: I am sorry to forget about this. update to recent release of sre-engine or import the source code here?

Comment (qingshi163) [reply vm/Cargo.toml:83]: Can you move the repr to rustpython-sre-engine.

**Comment by youknowone** [comment]
@qingshi163 other clippy errors looks mostly false positive or trivial. But could you check if `clippy::never_loop` one is intended or not?

File: vm/Cargo.toml (hunk 1 above)
Comment (youknowone) [reply vm/Cargo.toml]: I am going to import sre_engine in #5202

**Comment by qingshi163** [comment]
> @qingshi163 other clippy errors looks mostly false positive or trivial. But could you check if `clippy::never_loop` one is intended or not?

The loop of valued break is needed for performance issue, we just need to ignore the warning

File: vm/src/stdlib/sre.rs [hunk 2]
```diff
@@ -365,23 +414,48 @@ mod _sre {
...
+            // } else {
+            //     if is_template {
+            //         let re = vm.import("re", None, 0)?;
+            //         let func = re.get_attr("_compile_template", vm)?;
+            //         let filter = func.call((zelf.clone(), repl), vm)?;
+            //         (filter.is_callable(), filter)
+            //     } else {
+            //         (false, repl)
+            //     }
+            // };
```
Comment (youknowone) [inline vm/src/stdlib/sre.rs:458]:
```suggestion
```

File: vm/src/stdlib/sre.rs [hunk 3]
```diff
@@ -397,15 +471,36 @@ mod _sre {
...
+                        }
+                    };
+
+                    // if is_callable {
+                    //     let m = Match::new(&iter.state, zelf.clone(), string.clone());
+                    //     let ret = filter.call((m.into_ref(&vm.ctx),), vm)?;
+                    //     sublist.push(ret);
+                    // } else {
+                    //     sublist.push(filter.clone());
+                    // }
```
Comment (youknowone) [inline vm/src/stdlib/sre.rs:501]:
```suggestion
```

**Review by youknowone — APPROVED** [review]
Let's remove unused code blocks or add a TODO note to share why they are left as commented.

**Comment by youknowone** [comment]
Please revert the deletion once it is required

**Pull request merged**